drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1199 @@
+		LOG_ERROR("Invalid interface state");
+		return -EINVAL;
+	}
+	if (change_forbidden) {
+		interface_error_service(i2s,
+				"Failed to change interface state");
+		return -EIO;
+	}
+	i2s->state = new_state;
+	return 0;
+}
+
+static inline enum i2s_if_state interface_get_state(
+					struct i2s_nrfx_data *i2s)
+{
+	assert(i2s != NULL);
+	return i2s->state;
+}
+
+static int interface_restart(struct i2s_nrfx_data *i2s)
+{
+	return interface_set_state(i2s, I2S_IF_NEEDS_RESTART);
+}
+
+static int interface_stop(struct i2s_nrfx_data *i2s)
+{
+	int ret;
+
+	ret = interface_set_state(i2s, I2S_IF_STOPPING);
+	if (ret < 0) {
+		interface_error_service(i2s, "Failed to stop interface");
+		return ret;
+	}
+	return 0;
+}
+
+static int interface_stop_restart(struct i2s_nrfx_data *i2s,
+				struct channel_str *channel_to_stop_restart,
+				enum i2s_state other_channel_state)
+{
+	int ret;
+
+	ret = channel_change_state(channel_to_stop_restart,
+					I2S_STATE_STOPPING);
+	if (ret < 0) {
+		return ret;
+	}
+	if (other_channel_state == I2S_STATE_RUNNING) {
+		ret = interface_restart(i2s);
+		if (ret < 0) {
+			return ret;
+		}
+	} else {
+		ret = interface_stop(i2s);
+		if (ret < 0) {
+			interface_error_service(i2s,
+				"Failed to restart interface");
+			return ret;
+		}
+	}
+	return 0;
+}
+
+static int interface_start(struct i2s_nrfx_data *i2s)
+{
+	int ret;
+
+	ret = interface_set_state(i2s, I2S_IF_RUNNING);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/*nrfx_i2s_start() procedure needs buffer size in 32-bit word units*/
+	nrfx_err_t status = nrfx_i2s_start(&i2s->buffers,
+					   i2s->size / sizeof(u32_t), 0);
+
+	if (status != NRFX_SUCCESS) {
+		interface_error_service(i2s, "Failed to start interface");
+		ret = -EIO;
+	}
+
+	return ret;
+}
+
+static void interface_handler(nrfx_i2s_buffers_t const *p_released,
+				    u32_t status)
+{
+	struct i2s_nrfx_data *i2s = get_interface();
+	struct channel_str *rx_str = &i2s->channel_rx;
+	struct channel_str *tx_str = &i2s->channel_tx;
+	nrfx_i2s_buffers_t p_new_buffers;
+
+	p_new_buffers.p_rx_buffer = NULL;
+	p_new_buffers.p_tx_buffer = NULL;
+	if (rx_str->current_state != I2S_STATE_READY &&
+	    rx_str->current_state != I2S_STATE_NOT_READY) {
+		channel_rx_callback(i2s, p_released, status,
+					  &p_new_buffers);
+	}
+	if (tx_str->current_state != I2S_STATE_READY &&
+	    tx_str->current_state != I2S_STATE_NOT_READY) {
+		channel_tx_callback(i2s, p_released, status,
+					  &p_new_buffers);
+	}
+	if (next_buffers_needed(status)) {
+		if (interface_get_state(i2s) == I2S_IF_NEEDS_RESTART ||
+		    interface_get_state(i2s) == I2S_IF_STOPPING) {
+			nrfx_i2s_stop();
+			return;
+		} else if (interface_get_state(i2s) == I2S_IF_RUNNING) {
+			if (nrfx_i2s_next_buffers_set(&p_new_buffers) !=
+						NRFX_SUCCESS) {
+				nrfx_i2s_stop();
+				interface_error_service(i2s,
+						"Internal service error");
+				return;
+			}
+			if ((p_new_buffers.p_rx_buffer == NULL)
+			    && (p_new_buffers.p_tx_buffer == NULL)) {
```

Comment:
The logic used in this function should also be described a bit at least. For instance, when both these buffers my appear `NULL`?